kuksa_databroker/databroker/src/glob.rs

```diff
@@ -51,3 +52,113 @@ pub fn to_regex(glob: &str) -> Result<Regex, Error> {
     let re = to_regex_string(glob);
     Regex::new(&re).map_err(|_err| Error::RegexError)
 }
+
+fn has_asterisk_in_middle(input: &str) -> bool {
+    let input = input.trim(); // Remove leading and trailing whitespace
+    if input.starts_with('*') || input.ends_with('*') {
+        return false;
+    }
+
+    let asterisk_position = input.find('*');
+    if let Some(pos) = asterisk_position {
+        if input[pos - 1..].contains(".*.") {
+            return true;
+        }
+    }
+
+    false
+}
+
+pub fn to_regex_string_new(glob: &str) -> String {
+    let mut input = String::from(glob);
+
+    //check if asterisk in middle
+    if has_asterisk_in_middle(glob) {
+        input = glob.replace(".*.", ".");
+    }
+    // Construct regular expression
+
+    // Make sure we match the whole thing
+    // Start from the beginning
+    let mut re = String::from("^");
+
+    if input.ends_with(".*") {
+        re.push_str(input.replace(".*", ".[^.]+").as_str());
+    } else if input.ends_with(".**") {
+        re.push_str(input.replace(".**", ".*").as_str());
+    } else {
+        // Replace all "standalone" wildcards with ".*"
+        re.push_str(
+            &input
+                .split('.')
+                .map(|part| match part {
+                    "*" => "[^.]+",
+                    "**" => ".*",
+                    other => other,
+                })
+                .collect::<Vec<&str>>()
+                .join(r"\."),
+        );
+    }
+
+    // If it doesn't already end with a wildcard, add it
+    if !re.ends_with(".*") && !input.contains('*') {
+        re.push_str(".*");
+    }
+
+    // And finally, make sure we match until EOL
+    re.push('$');
+
+    re
+}
+
+pub fn to_regex_new(glob: &str) -> Result<Regex, Error> {
+    let re = to_regex_string_new(glob);
+    Regex::new(&re).map_err(|_err| Error::RegexError)
+}
+
+pub fn matches_path_pattern(input: &str) -> bool {
+    if input.starts_with("*.") {
+        return false;
+    }
```

Comment:
It was a bit hard to make a really generic regex that matches that case at this stage so that's why a decide to split for now into two regexes and this if clause. I do agree we need elegant regex solution that covers all use cases but since we change constantly opinions and perspectives, I didn't put to much effort into a good generic regex.